File: wpimath/src/test/native/cpp/kinematics/serde/ChassisSpeedsSerdeTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <gtest/gtest.h>

#include "frc/geometry/ChassisSpeeds.h"
#include "geometry2d.pb.h"

using namespace frc;

namespace {

using StructType = wpi::Struct<frc::ChassisSpeeds>;
using ProtoType = wpi::Protobuf<frc::ChassisSpeeds>;

constexpr std::array<uint8_t, StructType::kSize> create_test_buffer() {
  std::array<uint8_t, StructType::kSize> output;
  int buffer[] = {
    0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0,
};
  for (size_t idx = 0; idx < StructType::kSize; ++idx) {
    output[idx] = static_cast<uint8_t>(buffer[idx]);
  }
  return output;
}

std::array<uint8_t, StructType::kSize> kExpectedStructBuffer =
    create_test_buffer();

constexpr ChassisSpeeds kExpectedData{};
}  // namespace

TEST(ChassisSpeedsTest, StructPack) {
  uint8_t buffer[StructType::kSize];
  StructType::Pack(buffer, kExpectedData);

  for (size_t i = 0; i < StructType::kSize; ++i) {
    EXPECT_EQ(kExpectedStructBuffer[i], buffer[i]) << " on byte " << i;
  }
}

TEST(ChassisSpeedsTest, StructUnpack) {
  ChassisSpeeds unpacked_data = StructType::Unpack(kExpectedStructBuffer);

    EXPECT_EQ(kExpectedData.VxMps(), unpacked_data.VxMps());
    EXPECT_EQ(kExpectedData.VyMps(), unpacked_data.VyMps());
    EXPECT_EQ(kExpectedData.OmegaRps(), unpacked_data.OmegaRps());
}


TEST(ChassisSpeedsTest, ProtobufPack) {
  wpi::proto::ProtobufChassisSpeeds proto;
  ProtoType::Pack(&proto, kExpectedData);

    EXPECT_EQ(kExpectedData.VxMps(), proto.vx_mps());
    EXPECT_EQ(kExpectedData.VyMps(), proto.vy_mps());
    EXPECT_EQ(kExpectedData.OmegaRps(), proto.omega_rps());
}



TEST(ChassisSpeedsTest, ProtobufUnpack) {
  wpi::proto::ProtobufChassisSpeeds proto;
  proto.set_vx_mps(kExpectedData.VxMps());
  proto.set_vy_mps(kExpectedData.VyMps());
  proto.set_omega_rps(kExpectedData.OmegaRps());

  ChassisSpeeds unpacked_data = ProtoType::Unpack(proto);
    EXPECT_EQ(kExpectedData.VxMps(), unpacked_data.VxMps());
    EXPECT_EQ(kExpectedData.VyMps(), unpacked_data.VyMps());
    EXPECT_EQ(kExpectedData.OmegaRps(), unpacked_data.OmegaRps());
}

File: wpimath/src/test/java/edu/wpi/first/math/kinematics/serde/ChassisSpeedsSerdeTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.math.kinematics.serde;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;

import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.math.proto.Kinematics.ProtobufChassisSpeeds;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import org.junit.jupiter.api.Test;

public class ChassisSpeedsSerdeTest {
  private static final ChassisSpeeds DATA = new ChassisSpeeds(1.91, 2.29, 3.504);
  private static final byte[] STRUCT_BUFFER =
      new byte[] {
        -113, -62, -11, 40, 92, -113, -2, 63, 82, -72, 30, -123, -21, 81, 2, 64, -43, 120, -23, 38,
        49, 8, 12, 64
      };

  @Test
  void testStructPack() {
    ByteBuffer buffer = ByteBuffer.allocate(ChassisSpeeds.struct.getSize());
    buffer.order(ByteOrder.LITTLE_ENDIAN);
    ChassisSpeeds.struct.pack(buffer, DATA);

    byte[] actual = buffer.array();
    assertArrayEquals(actual, STRUCT_BUFFER);
  }

  @Test
  void testStructUnpack() {
    ByteBuffer buffer = ByteBuffer.wrap(STRUCT_BUFFER);
    buffer.order(ByteOrder.LITTLE_ENDIAN);

    ChassisSpeeds data = ChassisSpeeds.struct.unpack(buffer);
    assertEquals(DATA.vxMetersPerSecond, data.vxMetersPerSecond);
    assertEquals(DATA.vyMetersPerSecond, data.vyMetersPerSecond);
    assertEquals(DATA.omegaRadiansPerSecond, data.omegaRadiansPerSecond);
  }

  @Test
  void testProtoPack() {
    ProtobufChassisSpeeds proto = ChassisSpeeds.proto.createMessage();
    ChassisSpeeds.proto.pack(proto, DATA);

    assertEquals(DATA.vxMetersPerSecond, proto.getVxMps());
    assertEquals(DATA.vyMetersPerSecond, proto.getVyMps());
    assertEquals(DATA.omegaRadiansPerSecond, proto.getOmegaRps());
  }

  @Test
  void testProtoUnpack() {
    ProtobufChassisSpeeds proto = ChassisSpeeds.proto.createMessage();
    proto.setVxMps(DATA.vxMetersPerSecond);
    proto.setVyMps(DATA.vyMetersPerSecond);
    proto.setOmegaRps(DATA.omegaRadiansPerSecond);

    ChassisSpeeds data = ChassisSpeeds.proto.unpack(proto);
    assertEquals(DATA.vxMetersPerSecond, data.vxMetersPerSecond);
    assertEquals(DATA.vyMetersPerSecond, data.vyMetersPerSecond);
    assertEquals(DATA.omegaRadiansPerSecond, data.omegaRadiansPerSecond);
  }
}
